File: matlab/preprocessing/reinterleave_tiffs.m
function reinterleave_tiffs(A, source_dir, dest_dir,  split_channels)

    simeta = load(A.raw_simeta_path);

    if ~exist('split_channels', 'var')
        split_channels = false;
    end
    if split_channels
        fprintf('Splitting channels bec big TIFF.\n');
    end

    nslices = length(A.slices);
    nchannels = A.nchannels;
    nfiles = A.ntiffs;
    nvolumes = A.nvolumes;
    nframes = nslices*nvolumes*nchannels;
    d1=A.lines_per_frame; d2=A.pixels_per_line;

    sliceidxs = 1:nchannels:nslices*nchannels;
    tiffnames= dir(fullfile(source_dir, '*.tif'));
    tiffnames = {tiffnames(:).name}';
    fprintf('Found %i tiffs total for re-interleaving.\n', length(tiffnames));
    for fi=1:nfiles
         fprintf('d1: %i, d2: %s\n', d1, d2);
         newtiff = zeros(d1, d2, nframes);
         for sl = 1:nslices
             for ch=1:nchannels
                 suffix = sprintf('%s_Slice%02d_Channel%02d_File%03d.tif', A.base_filename, sl, ch, fi)
                 curr_tiff_idx = find(arrayfun(@(fn) length(strfind(tiffnames{fn}, suffix)), 1:length(tiffnames)))
                 
                 % This currently assumes only 2 channels...
                 currtiff = tiffnames{curr_tiff_idx};
                 currtiffpath = fullfile(source_dir, currtiff);
                 curr_file_name = sprintf('File%03d', fi);
                 if strfind(simeta.(curr_file_name).SI.VERSION_MAJOR, '2016') 
                     [tmp,~] = tiffRead(fullfile(source_dir,currtiff));
                 else
                     tmp = read_imgdata(currtiffpath);
                 end
                 if ch==1
                     newtiff(:,:,sliceidxs(sl):(nslices*nchannels):end) = tmp;
                 else
                     newtiff(:,:,(sliceidxs(sl)+1):(nslices*nchannels):end) = tmp;
                 end
             end
         end
         if split_channels
             for ch=1:nchannels
                newtiffname = sprintf('%s_File%03d_Channel%02d.tif', A.base_filename, fi, ch);
                tiffWrite(newtiff(:,:,ch:nchannels:end), newtiffname, dest_dir, 'int16');
             end 
         else
             newtiffname = sprintf('%s_File%03d.tif', A.base_filename, fi);
             tiffWrite(newtiff, newtiffname, dest_dir, 'int16');
         end
     end

end

File: python/preprocessing/reinterleave_tiffs.py
#!/usr/bin/env python2 
''' 
Run this script to interleave deinterleaved files. Assumes standard SliceNN_ChannelNN_FileNNN.tif format. 
Native format is int16. Loads reference .json file created in Step 1 of pipeline (process_raw.py)
 
Run python reinterleave_tiffs.py -h for all input options. 
''' 
import os 
import numpy as np 
import tifffile as tf 
from skimage import img_as_uint 
from skimage import exposure 
import optparse 
import shutil
import json
import re


def atoi(text):
    return int(text) if text.isdigit() else text

def natural_keys(text):
    return [ atoi(c) for c in re.split('(\d+)', text) ]


parser = optparse.OptionParser() 
parser.add_option('-S', '--source', action='store', dest='source', default='/nas/volume1/2photon/projects', help="source dir (parent directory of sessions) [default: '/nas/volume1/2photon/projects']")
parser.add_option('-E', '--experiment', action='store', dest='experiment', default='', help="stimulus-type (gratings_phaseMod, gratings_static, etc.)")
parser.add_option('-s', '--session', action='store', dest='session', default='', help='session dir (format: YYYMMDD_ANIMALID')
parser.add_option('-A', '--acq', action='store', dest='acquisition', default='', help="acquisition folder (ex: 'FOV1_zoom3x')")
parser.add_option('-f', '--functional', action='store', dest='functional_dir', default='functional', help="folder containing functional TIFFs. [default: 'functional']")
parser.add_option('-i', '--indir', action='store', dest='deinterleaved_dir', default='', help="folder containing tiffs to be interleaved")
parser.add_option('-o', '--outdir', action='store', dest='write_dir', default='', help="path to dir in which to write interleaved tiffs")


(options, args) = parser.parse_args()

source = options.source
experiment = options.experiment
session = options.session
acquisition = options.acquisition
functional_dir = options.functional_dir
deinterleaved_dir = options.deinterleaved_dir
write_dir = options.write_dir

# raw_simeta_basename = 'SI_raw_%s' % functional_dir
reference_info_fn = 'reference_%s.json' % functional_dir

acquisition_dir = os.path.join(source, experiment, session, acquisition)
data_dir =  os.path.join(acquisition_dir, functional_dir, 'DATA')
source_tiff_dir = os.path.join(data_dir, deinterleaved_dir)
if len(write_dir)==0:
    write_dir = data_dir
else:
    write_dir = os.path.join(data_dir, write_dir)
if not os.path.exists(write_dir):
    os.mkdir(write_dir)

tiffs = os.listdir(source_tiff_dir)
tiffs = sorted([t for t in tiffs if t.endswith('.tif')], key=natural_keys)
print "Found %i tiffs." % len(tiffs)

with open(os.path.join(acquisition_dir, reference_info_fn), 'r') as fr:
    ref = json.load(fr)

nfiles = ref['ntiffs']
nchannels = ref['nchannels']
nslices = len(ref['slices'])
nvolumes = ref['nvolumes']
ntotalframes = nslices*nvolumes*nchannels

sample = tf.imread(os.path.join(source_tiff_dir, tiffs[0]))
print sample.shape, sample.dtype

for fidx in range(nfiles):
    print "Interleaving File %i  f %i." % (int(fidx+1), nfiles)
    curr_file = "File%03d" % int(fidx+1)
    interleaved_tiff_fn = "{basename}_{currfile}.tif".format(basename=ref['base_filename'], currfile=curr_file)
    print "New tiff name:", interleaved_tiff_fn
    curr_file_fns = [t for t in tiffs if curr_file in t]
    print "Found %i tiffs for current file." % len(curr_file_fns)
    stack = np.empty((ntotalframes, sample.shape[1], sample.shape[2]), dtype=sample.dtype)
    for fn in curr_file_fns:
        curr_tiff = tf.imread(os.path.join(source_tiff_dir, fn))
        sl_idx = int(fn.split('Slice')[1][0:2]) - 1
        ch_idx = int(fn.split('Channel')[1][0:2]) - 1 
        slice_indices = np.arange((sl_idx*nchannels)+ch_idx, ntotalframes)
        idxs = slice_indices[::(nslices*nchannels)]
        stack[idxs,:,:] = curr_tiff
    
    tf.imsave(os.path.join(write_dir, interleaved_tiff_fn), stack)

